File: modules/file.go
package modules

import (
	"os"
	"path"
)

// ---------------------------------
// Vars

// FileStruct struct for the project
type FileStruct struct {
	Src    string `toml:"source"`
	Dest   string `toml:"destination"`
	Ignore string
	Force  bool
	Order  int
	Env    string
	Sys    string
}

// ---------------------------------
// Public functions

// FileTask for init
func FileTask(config []FileStruct, taskType string, order int, env string, sys string) {
	for _, task := range config {
		var shouldContinue bool
		if task.Order, task.Env, task.Sys, shouldContinue = InitDecision(
			task.Order, task.Env, task.Sys, order, env, sys,
		); shouldContinue {
			continue
		}

		// Get the right paths
		src, ignore := GetPaths(task.Src, task.Ignore)
		originalDest := task.Dest

		// Go through each in the glob
		for _, file := range src {
			// Check if is in the ignore
			if ArrContainsStr(ignore, file) {
				continue
			}

			Log(taskType, file)

			// Reset dest
			task.Dest = originalDest

			// Needs this
			task.Dest = ConstructDest(taskType, task.Dest, file, task.Src)
			task.Src = file

			// Instantiate log
			var logVal string
			var err error

			switch taskType {
			case "rename":
				logVal, err = FileRename(task)
			case "remove":
				logVal, err = FileRemove(task)
			default:
				logVal, err = FileCopy(task)
			}

			// Log whatever
			LogErr(taskType, err)
			Log(taskType+"] [result", logVal)
		}
	}
}

// FileCopy copies * from source to destination
func FileCopy(file FileStruct) (log string, err error) {
	// Remove the file if it has force
	if !NotExist(file.Dest) && file.Force {
		FileRemove(FileStruct{Src: file.Dest})
	}

	srcFile, srcErr := os.Stat(file.Src)

	switch {
	// Check if src exists
	case srcErr != nil:
		err = srcErr
	// Check if src is directory
	case srcFile.IsDir():
		// TODO: What about the ignores?
		err = folderCopy(file)
	// It must be a file
	default:
		err = fileCopy(file)
	}

	return log, err
}

// FileRemove removes * from source to destination
func FileRemove(file FileStruct) (log string, err error) {
	src := file.Src

	err = os.RemoveAll(src)
	if err != nil {
		return "", err
	}

	return
}

// FileRename renames * from source to destination
func FileRename(file FileStruct) (log string, err error) {
	src := file.Src
	dest := file.Dest

	err = os.Rename(src, dest)
	if err != nil {
		return "", err
	}

	return
}

// ---------------------------------
// Private functions

// Copy file from source to destination
func fileCopy(file FileStruct) (err error) {
	src := file.Src
	dest := file.Dest

	// We need to ensure that the folder exists
	err = EnsurePath(dest)
	if err != nil {
		return err
	}

	// Remove old destination file if exists
	FileRemove(FileStruct{Src: dest})

	// Copy the file
	err = os.Link(src, dest)
	if err != nil {
		return err
	}

	return
}

// Copy folder from source to destination
func folderCopy(file FileStruct) (err error) {
	src := file.Src
	dest := file.Dest

	// We need to ensure that the folder exists
	err = EnsurePath(dest)
	if err != nil {
		return err
	}

	// Remove old destination file if exists
	FileRemove(FileStruct{Src: dest})

	// Make folder
	mode := os.FileMode(int(0777))
	if err := os.MkdirAll(dest, mode); err != nil {
		return err
	}

	// Read the files inside
	dir, _ := os.Open(src)
	files, _ := dir.Readdir(-1)

	// Loop each file
	for _, file := range files {
		srcPath := path.Join(src, file.Name())
		destPath := path.Join(dest, file.Name())

		// Copy new file
		_, err = FileCopy(FileStruct{Src: srcPath, Dest: destPath})
		if err != nil {
			return err
		}
	}

	return
}

File: modules/file.js
'use strict';

// -----------------------------------------
// IMPORTS

var fs = require('fs');
var path = require('path');
var exec = require('sync-exec');
var Joi = require('joi');

var tools = require('./tools/main');
var validate = tools.validate;

// -----------------------------------------
// VARS

var struct = Joi.object().keys({
    src: Joi.string().required(), // `toml:"source"`
    dest: Joi.string().optional(), // `toml:"destination"`
    ignore: Joi.string().default('').allow(''),
    order: Joi.number().default(0),
    env: Joi.string().allow('').default(''),
    cmd: Joi.string().allow('').default(''),
    sys: Joi.string().allow('').default('all')
});

// -----------------------------------------
// PUBLIC FUNCTIONS

/**
 * Create task for init
 * @param  {object} bedrockObj
 * @param  {object} config
 * @param  {string} taskType
 */
function task(bedrockObj, config, taskType) {
    validate.type(
        { config: config, taskType: taskType },
        { config: Joi.array().items(struct), taskType: Joi.string() }
    );

    // Go through each task
    config.forEach(function (configTask) {
        var shouldContinue = tools.decide(bedrockObj, configTask);
        var ignore;
        var src;

        if (shouldContinue) {
            return;
        }

        // Get the right paths
        src = tools.getGlob(configTask.src, taskType === 'remove', taskType === 'remove');
        ignore = configTask.ignore ? tools.getGlob(configTask.ignore) : [];

        // Lets filter files
        ignore = ignore.map(function (file) {
            return file.relative;
        });
        src = src.filter(function (file) {
            return !tools.arrContainsStr(ignore, file.relative);
        });

        // Go through each in the glob
        src.forEach(function (file) {
            // Create task
            var newTask = { src: file.absolute };
            var dest;

            tools.log(file.relative);

            if (taskType !== 'remove') {
                dest = path.join(configTask.dest, file.relative);
                newTask.dest = tools.getAbsolute(dest);
            }

            switch (taskType) {
            case 'rename':
                rename(newTask);
                break;
            case 'remove':
                remove(newTask);
                break;
            default:
                copy(newTask);
            }
        });
    });
}

/**
 * Copies * from source to destination
 * @param  {object} file
 */
function copy(file) {
    var data;

    validate.type({ file: file }, { file: struct });

    if (tools.notExist(file.src)) {
        return tools.logErr('File doesn\'t exist');
    }

    if (tools.isDirectory(file.src)) {
        // We should recursive in case of directory
        task({
            order: file.order,
            env: file.env,
            cmd: file.cmd,
            sys: file.sys
        }, [{
            src: file.src + '/**/*',
            dest: file.dest,
            order: file.order,
            env: file.env,
            cmd: file.cmd,
            sys: file.sys
        }], 'copy');
    } else {
        // First ensure the path
        tools.ensurePath(file.dest);

        // Remove file if exists
        if (!tools.notExist(file.dest)) {
            remove({ src: file.dest });
        }

        // Now copy the file
        data = exec('cp ' + file.src + ' ' + file.dest);

        if (data.stderr && data.stderr.length) {
            tools.logErr(data.stderr);
        } else if (data.stdout && data.stdout.length) {
            tools.log(data.stdout);
        }
    }
}

/**
 * Removes * from source
 * @param  {object} file
 */
function remove(file) {
    var data;
    var src;

    validate.type({ file: file }, { file: struct });

    src = file.src.replace(/ /g, '');
    if (src === '.' || src === '' || src === '/') {
        return tools.logErr('Trying to remove a global directory!');
    } else if (tools.notExist(src)) {
        return;
    }

    data = exec('rm -rf ' + src);
    if (data.stderr && data.stderr.length) {
        tools.logErr(data.stderr);
    } else if (data.stdout && data.stdout.length) {
        tools.log(data.stdout);
    }
}

/**
 * Renames * from source to destination
 * @param  {object} file
 */
function rename(file) {
    validate.type({ file: file }, { file: struct });

    if (tools.notExist(file.src)) {
        return;
    }

    fs.renameSync(file.src, file.dest);
}

// -----------------------------------------
// PRIVATE FUNCTIONS

// -----------------------------------------
// EXPORTS

module.exports = {
    struct: struct,
    task: task,
    copy: copy,
    remove: remove,
    rename: rename
};
